### PR TITLE
abstract away document collection

### PR DESCRIPTION
**Pull request by zhangxin511**
## Description

Abstract away `document` collection usage by replace with `documentRepository`, which would used for customization override

## Reviewer Guidance

Please have a double eye on the refactored code to make sure there is no wrong changes, thank you for the helps!

### CODE REVIEW

File: server/routerlicious/packages/memory-orderer/src/localOrdererSetup.ts [hunk 1]
```diff
@@ -31,11 +33,18 @@ export class LocalOrdererSetup implements ILocalOrdererSetup {
 		return this.storage.getOrCreateDocument(this.tenantId, this.documentId);
 	}
 
+	/**
+	 * @deprecated use documentRepositoryP() instead
+	 */
 	// eslint-disable-next-line @typescript-eslint/promise-function-async
 	public documentCollectionP(): Promise<ICollection<IDocument>> {
 		return this.databaseManager.getDocumentCollection();
 	}
```
Comment (yangg-msft) [inline server/routerlicious/packages/memory-orderer/src/localOrdererSetup.ts:42]: Let's emit a log to indicate this API is still being exercised?

File: server/routerlicious/packages/routerlicious/src/scriptorium/index.ts [hunk 2]
```diff
@@ -79,23 +48,5 @@ export async function create(config: Provider): Promise<IPartitionLambdaFactory>
 			: opCollection.createTTLIndex({ mongoTimestamp: 1 }, mongoExpireAfterSeconds));
 	}
 
-	executeOnInterval(
-		async () =>
```
Comment (yangg-msft) [inline server/routerlicious/packages/routerlicious/src/scriptorium/index.ts:83]: I understand we don't exercise this feature, but other R11s consumers may need it?

Comment (zhangxin511) [reply server/routerlicious/packages/routerlicious/src/scriptorium/index.ts:83]: What is other R11s consumer you are refering @yangg-msft ? Real customer or code consumer? If code consumer, then there is no other consumer, if it is a concern for real customer, then I can add it back

File: server/routerlicious/packages/services-core/src/database.ts [hunk 3]
```diff
@@ -38,6 +38,47 @@ export interface IDatabaseManager {
 	): Promise<ICollection<ISequencedOperationMessage>>;
 }
 
+/**
+ * Abstract away IDocument collection logics and
```
Comment (pradeepvairamani) [inline server/routerlicious/packages/services-core/src/database.ts:42]: Nit: typo

File: server/routerlicious/packages/routerlicious/src/scriptorium/index.ts (hunk 2 above)
Comment (zhangxin511) [reply server/routerlicious/packages/routerlicious/src/scriptorium/index.ts:83]: Readded this piece of code. Changed the execution more reasonable way but it still won't work

File: server/routerlicious/packages/services-core/src/database.ts (hunk 3 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services-core/src/database.ts]: Addressed

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts [hunk 4]
```diff
@@ -123,7 +123,10 @@ export class AlfredResources implements core.IResources {
 }
 
 export class AlfredResourcesFactory implements core.IResourcesFactory<AlfredResources> {
-	public async create(config: Provider): Promise<AlfredResources> {
+	public async create(
+		config: Provider,
+		customization?: Record<string, any>,
```
Comment (hedasilv) [inline server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:128]: Do we plan to support other scenarios than customizing the `documentRepository`? Is that why we wrap all that into this customization object? If yes, would `customizations` be a better name?

File: server/routerlicious/packages/routerlicious/src/deli/index.ts [hunk 5]
```diff
@@ -38,7 +41,6 @@ export async function deliCreate(config: Provider): Promise<core.IPartitionLambd
 	const authEndpoint = config.get("auth:endpoint");
 	const internalHistorianUrl = config.get("worker:internalBlobStorageUrl");
 	const tenantManager = new services.TenantManager(authEndpoint, internalHistorianUrl);
-	const globalDbEnabled = config.get("mongo:globalDbEnabled") as boolean;
```
[the remark this answers is not in the document]
Comment (zhangxin511) [reply server/routerlicious/packages/routerlicious/src/deli/index.ts:41]: I can revert it back,  I thought it is just something we used

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts (hunk 4 above)
Comment (zhangxin511) [reply server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts]: Addressed

File: server/routerlicious/packages/test-utils/src/testNonImplementDocumentRepository.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IDocument, IDocumentRepository } from "@fluidframework/server-services-core";
+
+const defaultErrorMsg = "Method not implemented. Provide your own mock.";
+export class TestNonImplementDocumentRepository implements IDocumentRepository {
```
[the remark this answers is not in the document]
Comment (zhangxin511) [reply server/routerlicious/packages/test-utils/src/testNonImplementDocumentRepository.ts]: addressed

File: server/routerlicious/packages/services/src/mongodb.ts [hunk 7]
```diff
@@ -71,8 +71,8 @@ export class MongoCollection<T> implements core.ICollection<T>, core.IRetryable
 		return this.requestWithRetry(req, "MongoCollection.find", query);
 	}
 
-	public async findOne(query: object): Promise<T> {
-		const req: () => Promise<T> = async () => this.collection.findOne(query);
+	public async findOne(query: object, option?: FindOneOptions): Promise<T> {
```
[the remark this answers is not in the document]
Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts]: addressed

**Review by hedasilv — APPROVED** [review]

**Review by pradeepvairamani — APPROVED** [review]

**Review by arikt-ms — APPROVED** [review]

**Review by yunho-microsoft — APPROVED** [review]

File: server/routerlicious/packages/services-core/src/mongoDocumentRepository.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+ * Licensed under the MIT License.
+ */
+
+import { ICollection, IDb, IDocumentRepository } from "./database";
+import { IDocument } from "./document";
+
+export class MongoDocumentRepository implements IDocumentRepository {
+	constructor(private readonly collection: ICollection<IDocument>) {}
+
+	public static create(database: IDb, collectionName: string): MongoDocumentRepository {
```
[the remark this answers is not in the document]
Comment (zhangxin511) [reply server/routerlicious/packages/services-core/src/mongoDocumentRepository.ts]: It was created at beginning of the phase when I tried to eliminate the `ICollection<IDocument>` usage, but It seems impossible to remove all these

**Pull request merged**